Implement Metadata store (WP2 for QEP-91) (part 1)

nyalldawson opened the pull request:
## Description
This PR partly implements Work Package 2 from QEP-91 - implementing a metadata store in QGIS.

It implements a new class "QgsLayerMetadata". QgsLayerMetadata handles storage and management of the metadata for a QgsMapLayer. This class is an internal QGIS format with a common metadata structure, which allows for code to access the metadata properties for layers in a uniform way.

The metadata store is designed to be compatible with the Dublin Core metadata specifications, and will be expanded to allow compatibility with ISO specifications in future releases. However, the QGIS internal schema does not represent a superset of all existing metadata schemas and accordingly conversion from specific metadata formats to QgsLayerMetadata may result in a loss of information.

This class follows the schema defined in https://github.com/qgis/QGIS/pull/4330. #4330 is a prerequisite before this can be merged.

Note 1: please do not comment on the metadata schema here - those comments must be directed to #4330. Feedback on this PR should be limited to the API/code only.

Note 2: this PR does not implement the extent component of the schema - that is blocked by #4349 and #4353.

Note 3: this PR does not implement persistent storage and parsing of metadata. That will be implemented in a follow up package as part of QEP-91. To workaround this limitation, I've temporarily added (limited) support for storing QgsLayerMetadata in a layer's custom properties, allowing persistence of the metadata within a single project. **I consider this a temporary hack, and this support WILL be removed when the proper storage handling component of QEP-91 is implemented**

Note 4: The redundant API calls from QgsMapLayer (eg QgsMapLayer::abstract(), keywordList(), attribution(), title() ) will be removed following merge of this PR.

## Checklist

> Reviewing is a process done by magic fairies, mostly on a bribery basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

nyalldawson committed:
Preparation for metadata directory

nyalldawson committed:
Remove duplicate gui/effects include

nyalldawson committed:
Metadata class shell

nyalldawson committed:
Attach layer metadata to QgsMapLayer

nyalldawson committed:
Implement metadata store for map layers

QgsLayerMetadata handles storage and management of the metadata
for a QgsMapLayer. This class is an internal QGIS format with a common
metadata structure, which allows for code to access the metadata properties for
layers in a uniform way.

The metadata store is designed to be compatible with the Dublin Core metadata
specifications, and will be expanded to allow compatibility with ISO specifications
in future releases. However, the QGIS internal schema does not represent a superset
of all existing metadata schemas and accordingly conversion from specific
metadata formats to QgsLayerMetadata may result in a loss of information.

This class is designed to follow the specifications detailed in
the schema definition available at resources/qgis-resource-metadata.xsd
within the QGIS source code.

nyalldawson committed:
Add temporary methods to save layer metadata in layer custom properties

Allows temporary persistance of metadata inside a single project. Will
be removed when later work packages from the metadata project are
merged.

nyalldawson commented:
@tomkralidis  @kalxas Can you please review and validate the API docs here? I want to make sure that the descriptions for the various elements are accurate.

luipir commented:
@nyalldawson I'm starting to be confused, there is a relation of this PR with grant proposal ?
"IMPLEMENT QGIS METADATA STORAGE SUPPORT" listed in:
https://qgisblog.files.wordpress.com/2017/04/qgisgrantapplicationssummary-march2017.pdf

nyalldawson commented:
> @nyalldawson I'm starting to be confused, there is a relation of this PR with grant proposal ?
"IMPLEMENT QGIS METADATA STORAGE SUPPORT" listed in:
https://qgisblog.files.wordpress.com/2017/04/qgisgrantapplicationssummary-march2017.pdf

Yes - there is. The overall project is split into multiple chunks, and you can find a summary of these at https://github.com/qgis/QGIS-Enhancement-Proposals/issues/91

Part 1 is #4330
Part 2 is this one
Part 3 is implementing the real storage support for metadata (not the layer custom property hack used here), which is what that grant proposal is for

(+ there's further parts listed in the QEP)

luipir commented:
ok, thanks for the explanation, I didn't follow in detail

tomkralidis commented:
@nyalldawson looks good.  Note I've added `rm:license` to the schema in https://github.com/kalxas/QGIS/commit/b8a6965475be303efda54b1fd7bc00ba7d2957be FYI.

timlinux commented on tests/src/python/test_qgslayermetadata.py:
```diff
@@ -0,0 +1,305 @@
+        c.organization = 'Bergen co'
+        c.position = 'prince'
+        c.voice = '1500 515 555'
+        c.fax = 'who the f*** still uses fax?'
```
:-)

kalxas replied:
lol

timlinux commented:
Lovely stuff @nyalldawson and definitely one of the most entertaining unit tests I have read :-)

timlinux commented:
@nyalldawson are you thinking of moving straight on to merge in QGIS 3.0 or do you still want to stage it in our fork?

nyalldawson commented:
@tomkralidis 
> Note I've added rm:license to the schema in kalxas@b8a6965 FYI.

Thanks - I'll keep and eye on #4330 and keep this PR in sync

nyalldawson commented:
@timlinux 
> are you thinking of moving straight on to merge in QGIS 3.0 or do you still want to stage it in our fork?

As it turns out I'd be happy to merge this to master. I think in the end the changes are self-contained enough and the addition of the (albeit temporary) project saving support means that this is still a functional change on its own. What do you think?

nyalldawson commented:
> Lovely stuff @nyalldawson and definitely one of the most entertaining unit tests I have read :-)

I'm glad someone noticed ;) I was getting rather bored entering all those text strings...

kalxas commented:
Looks good to me too.
Should we merge #4330 before this one?

nyalldawson committed:
Add metadata validator

Adds a new abstract base class QgsMetadataValidator for validating
metadata against standard schemas.

Initially only QgsNativeMetadataValidator for validating against
the native QGIS metadata schema is implemented.

In future this could be extended with Dublin Core, ISO 19115
validators, etc...

nyalldawson committed:
Add license section to metadata

nyalldawson committed:
Add history item to metadata

nyalldawson committed:
Add doxygen note about metadata validation

nyalldawson committed:
Add spatial and temporal extents to metadata

nyalldawson commented:
This now completely encapsulates the schema from #4330. 

Assuming no further changes to that PR this is ready to go!

timlinux commented:
@nyalldawson can you merge this?

nyalldawson commented:
@timlinux shouldn't merge #4330 first?

nyalldawson committed:
Sipify update

Gustry commented on src/core/metadata/qgslayermetadata.cpp:
```diff
@@ -0,0 +1,287 @@
+  layer->setCustomProperty( QStringLiteral( "metadata/constraints" ), QVariant::fromValue( mConstraints ) );
+  layer->setCustomProperty( QStringLiteral( "metadata/keywords" ), QVariant::fromValue( mKeywords ) );
+  layer->setCustomProperty( QStringLiteral( "metadata/contacts" ), QVariant::fromValue( mContacts ) );
+  layer->setCustomProperty( QStringLiteral( "metadata/links" ), QVariant::fromValue( mLinks ) );
```
AFAIK, all these structs `links`,  `extent`,  `keywords` ... are not saved in the QGIS Project. Is-it expected @nyalldawson? It works for simple QString or QStringList in the XML.
